Deprecate old fft functions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45164 torch.fft: Two dimensional FFT functions
* #45163 codegen: Improve array default handing
* #44877 torch.fft helper functions
* **#44876 Deprecate old fft functions**
* #44550 torch.fft: Multi-dimensional transforms

Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

peterbell10 committed:
Deprecate old fft functions

[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -561,13 +561,19 @@ void _cufft_clear_plan_cache(int64_t device_index) {
 
 Tensor fft(const Tensor& self, const int64_t signal_ndim, const bool normalized) {
+  TORCH_WARN_ONCE(
+    "torch.fft is deprecated and will be removed in pytorch 1.8. "
```
Reviewing this now because I'd like to get this in for 1.7 (branch cut next week Tuesday) so `torch.fft()` is deprecated in 1.7 and can be removed in 1.8. Realistically this PR and the one it's stacked will need to be done and landable before Monday to make that happen. Let me know if there's something I can do to help, @peterbell10. 

Overall this looks good but it needs two things:

- the _tensor_docs.py for `torch.fft` also need a (deprecation) warning with language like this (except it can link to the torch.fft module documentation)
- this warning should tell people to import torch.fft to get access to the functions that are replacing `torch.fft()`

Maybe something like:

"The function torch.fft is deprecated and will be removed in a future PyTorch release. Use the new torch.fft module functions, instead, by importing torch.fft and calling torch.fft.fft or torch.fft.fftn." 

?

peterbell10 replied:
@mruberry I've updated the docs and warning message. Is there a policy on giving the version a deprecated function will be removed? I've put `torch.fft()` as expiring in 1.8 but left the others as "a furture PyTorch release" since that's not a priority.

mruberry replied:
No policy. "In a future PyTorch release" gives us some latitude.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -576,6 +582,9 @@ Tensor ifft(const Tensor& self, const int64_t signal_ndim, const bool normalized
 
 Tensor rfft(const Tensor& self, const int64_t signal_ndim, const bool normalized,
             const bool onesided) {
+  TORCH_WARN_ONCE(
```
Deprecating `torch.fft()` is the priority (since its name conflicts with the module), but deprecating these is OK if the same pattern is followed. I.e.:

""The function torch.rfft is deprecated and will be removed in a future PyTorch release. Use the new torch.fft module functions, instead, by importing torch.fft and calling torch.fft.rfft or torch.fft.rfftn."

And also updating the torch.rfft and torch.irfft docs.

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

peterbell10 commented on torch/_torch_docs.py:
```diff
@@ -8489,6 +8489,12 @@ def merge_dicts(*dicts):
+.. deprecated:: 1.7.0
+    The function :func:`torch.fft` is deprecated and will be removed in a
+    PyTorch 1.8. Use the new :mod:`torch.fft` module functions instead, by
+    importing :mod:`torch.fft` and calling :func:`torch.fft.fft` or
```
I'll note that `torch.fft` unfortunately links to the function instead of the module. I was hoping `:mod:` would disambiguate, but I guess it's unlikely to work properly until the function is removed.

mruberry replied:
Darn. I guess we can just write the url by hand for now? It's very unlikely to change.

peterbell10 replied:
Used a reference so I didn't have to hard-code the URL.

mruberry replied:
Smart solution.

peterbell10 committed:
Update on "Deprecate old fft functions"




[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -561,13 +561,21 @@ void _cufft_clear_plan_cache(int64_t device_index) {
 Tensor fft(const Tensor& self, const int64_t signal_ndim, const bool normalized) {
+  TORCH_WARN_ONCE(
+    "The function torch.fft is deprecated and will be removed in PyTorch 1.8. "
+    "Use the new torch.fft module function instead, by importing torch.fft and "
```
"module function" -> "module functions, instead"

here and below

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -576,6 +584,10 @@ Tensor ifft(const Tensor& self, const int64_t signal_ndim, const bool normalized
             const bool onesided) {
+  TORCH_WARN_ONCE(
+    "The function torch.rfft is deprecated and will be removed in a future "
+    "PyTorch realease. Use the new torch.fft module function instead, by "
```
"realease" -> "release"

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -576,6 +584,10 @@ Tensor ifft(const Tensor& self, const int64_t signal_ndim, const bool normalized
+  TORCH_WARN_ONCE(
+    "The function torch.rfft is deprecated and will be removed in a future "
+    "PyTorch realease. Use the new torch.fft module function instead, by "
+    "importing torch.fft and calling torch.fft.fft or torch.fft.rfft.");
```
Should this be rfft or rfftn?

mruberry reviewed (approved):
Nice work, @peterbell10. Just a few minor corrections to be applied. Ping me when this is ready.

peterbell10 committed:
Update on "Deprecate old fft functions"


Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"


Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"


Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

[ghstack-poisoned]

peterbell10 commented:
@mruberry I've corrected the warnings.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8514,6 +8514,12 @@ def merge_dicts(*dicts):
 The inverse of this function is :func:`~torch.ifft`.
 
+.. deprecated:: 1.7.0
+    The function :func:`torch.fft` is deprecated and will be removed in a
```
"in a PyTorch 1.8" -> "in PyTorch 1.8"

peterbell10 replied:
@mruberry fixed.

peterbell10 committed:
Update on "Deprecate old fft functions"


Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

[ghstack-poisoned]

peterbell10 committed:
Update on "Deprecate old fft functions"


Differential Revision: [D23866715](https://our.internmc.facebook.com/intern/diff/D23866715)

[ghstack-poisoned]

mruberry reviewed (approved):
Awesome! Thanks @peterbell10!

fritzo commented:
Note the 1.7 release notes state

> Please use torch.fft.foo as a drop-in replacement for torch.foo for the following functions: fft, ifft, rfft and irfft.

however the `rfft` and `irfft` signatures have changed. Would PyTorch devs have time to update release notes with old-versus-new translations for these changed functions?

mruberry commented:
> Note the 1.7 release notes state
> 
> > Please use torch.fft.foo as a drop-in replacement for torch.foo for the following functions: fft, ifft, rfft and irfft.
> 
> however the `rfft` and `irfft` signatures have changed. Would PyTorch devs have time to update release notes with old-versus-new translations for these changed functions?

The release notes are, unfortunately, incorrect. All of these functions are different from each other, as described in the Wiki article here: https://github.com/pytorch/pytorch/wiki/The-torch.fft-module-in-PyTorch-1.7. You can translate rfft and irfft calls to rfftn and irfftn calls, however.
